ev3dev/control/GyroBalancer.py

```diff
@@ -1,398 +1,477 @@
-#
-# The above copyright notice and this permission notice shall be included in all
-# copies or substantial portions of the Software.
-#
-# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-# SOFTWARE.
+"""Module for a robot that stands on two wheels and uses a gyro sensor.
+
+The robot (eg. BALANC3R) will to keep its balance and move in response to
+the remote control. This code was adapted from Laurens Valk's script at
+https://github.com/laurensvalk/segway.
 
-"""
-This is a class-based version of https://github.com/laurensvalk/segway
 """
 import logging
-import math
 import time
+import sys
 from collections import deque
-from ev3dev.auto import *
-from ev3dev.helper import LargeMotorPair
-
+from ev3dev.power import PowerSupply
+from ev3dev.motor import LargeMotor, OUTPUT_A, OUTPUT_D
+from ev3dev.sensor.lego import GyroSensor, InfraredSensor, TouchSensor
 
 log = logging.getLogger(__name__)
 
-
 ########################################################################
-##
-## File I/O functions
-##
+# File I/O functions
 ########################################################################
 
-# Function for fast reading from sensor files
+
 def FastRead(infile):
+    """Function for fast reading from sensor files."""
     infile.seek(0)
-    return int(infile.read().decode().strip())
+    return(int(infile.read().decode().strip()))
 
 
-# Function for fast writing to motor files
 def FastWrite(outfile, value):
+    """Function for fast writing to motor files."""
     outfile.truncate(0)
     outfile.write(str(int(value)))
     outfile.flush()
 
 
-# Function to set the duty cycle of the motors
-def SetDuty(motorDutyFileHandle, duty):
-    # Clamp the value between -100 and 100
-    duty = min(max(duty, -100), 100)
+def eprint(*args, **kwargs):
+    """Debug print (https://stackoverflow.com/a/14981125)."""
+    print(*args, file=sys.stderr, **kwargs)
+
+
+def SetDuty(motorDutyFileHandle, duty, frictionOffset, voltageCompensation):
+    """Function to set the duty cycle of the motors."""
+    # Compensate for nominal voltage and round the input
+    dutyInt = int(round(duty*voltageCompensation))
+
+    # Add or subtract offset and clamp the value between -100 and 100
+    if dutyInt > 0:
+        dutyInt = min(100, dutyInt + frictionOffset)
+    elif dutyInt < 0:
+        dutyInt = max(-100, dutyInt - frictionOffset)
 
     # Apply the signal to the motor
-    FastWrite(motorDutyFileHandle, duty)
+    FastWrite(motorDutyFileHandle, dutyInt)
 
 
-class GyroBalancer(LargeMotorPair):
+class GyroBalancer():
```

Comment:
Would have it in inherit from `object`

Reply:
Even though Python 3 does not require it ? I thought that was only important for code that may be run under Python 2.